ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java

```diff
@@ -0,0 +1,367 @@
+          commandLog);
+      this.ksqlResource = new KsqlResource(
+          ksqlConfig,
+          ksqlEngine,
+          fakeCommandQueue,
+          0
+      );
+      this.statementExecutor = new StatementExecutor(
+          ksqlConfig,
+          ksqlEngine,
+          new StatementParser(ksqlEngine));
+    }
+
+    void recover() {
+      this.statementExecutor.handleRestoration(this.fakeCommandQueue.getRestoreCommands());
+    }
+
+    void executeCommands() {
+      fakeCommandQueue.getNewCommands().forEach(
+          c -> statementExecutor.handleStatement(
+              c.getCommand(), c.getCommandId(), Optional.empty())
+      );
+    }
+
+    void submitCommands(final String ...statements) {
+      for (final String statement : statements) {
+        final Response response = ksqlResource.handleKsqlStatements(
+            new KsqlRequest(statement, Collections.emptyMap()));
+        assertThat(response.getStatus(), equalTo(200));
+        executeCommands();
+      }
+    }
+  }
+
+  private void assertTopicsEqual(final KsqlTopic topic, final KsqlTopic otherTopic) {
+    assertThat(topic.getTopicName(), equalTo(otherTopic.getTopicName()));
+    assertThat(topic.getKafkaTopicName(), equalTo(otherTopic.getKafkaTopicName()));
+    assertThat(
+        topic.getKsqlTopicSerDe().getClass(),
+        equalTo(otherTopic.getKsqlTopicSerDe().getClass()));
+  }
+
+  private void assertTablesEqual(final KsqlTable table, final StructuredDataSource other) {
+    assertThat(other, instanceOf(KsqlTable.class));
+    final KsqlTable otherTable = (KsqlTable) other;
+    assertThat(table.isWindowed(), equalTo(otherTable.isWindowed()));
+  }
+
+  private void assertSourcesEqual(
+      final StructuredDataSource source,
+      final StructuredDataSource other) {
+    assertThat(source.getClass(), equalTo(other.getClass()));
+    assertThat(source.getName(), equalTo(other.getName()));
+    assertThat(source.getDataSourceType(), equalTo(other.getDataSourceType()));
+    assertThat(source.getSchema(), equalTo(other.getSchema()));
+    assertThat(source.getSqlExpression(), equalTo(other.getSqlExpression()));
+    assertThat(
+        source.getTimestampExtractionPolicy(), equalTo(other.getTimestampExtractionPolicy()));
+    assertThat(source, anyOf(instanceOf(KsqlStream.class), instanceOf(KsqlTable.class)));
+    assertTopicsEqual(source.getKsqlTopic(), other.getKsqlTopic());
+    if (source instanceof KsqlTable) {
+      assertTablesEqual((KsqlTable) source, other);
+    }
+  }
+
+  private void shouldRecover(final List<QueuedCommand> commands) {
+    // Given:
+    final KsqlServer executeServer = new KsqlServer(commands);
+    executeServer.executeCommands();
+    final KsqlEngine engine = executeServer.ksqlEngine;
+
+    // When:
+    final KsqlServer recoverServer = new KsqlServer(commands);
+    recoverServer.recover();
+    final KsqlEngine recovered = recoverServer.ksqlEngine;
+
+    // Then:
+    assertThat(
```

Comment:
This was intentional. It doesn't really make sense semantically to compare meta stores outside of tests. When would it ever make sense do this in real ksql code?

Reply:
Humm... good point

You could use the `hashCode` to check if different servers were in sink. 

You could use `equals` to check is a statement changed the metastore. 

Etc.... 

Though... playing devils advocate, we don't do either of those things and I'm also not a fan of production code that is also only used by tests. 

My main reason for suggesting this was to make the test more readable / shorter, as its quite long. The other way this can be achieved is using a custom matcher. Do you have any experience with this?

```
assertThat(recovered.getMetaStore(), is(metaStore(engine.getMetaStore())));
```

Where `metaStore()` is:

```
private static Matcher<? super MetaStore> metaStore(final MetaStore expected) {
   final Matcher<Set<String>> dataSourcesMatcher = Matchers.is(expected.getAllStructuredDataSourceNames());
    ...
    
    return new TypeSafeDiagnosingMatcher<MetaStore>() {

      @Override
      protected boolean matchesSafely(final MetaStore actual, final Description description) {
        if (!dataSourcesMatcher.matches(actual.getAllStructuredDataSourceNames())) {
          description.appendText("data source names mismatch: ");
          dataSourcesMatcher.describeMismatch(actual.getAllStructuredDataSourceNames(), description);
          return false;
        }
        
        ...
      
        return true;
      }

      @Override
      public void describeTo(final Description description) {
          description.appendText("with data source names: ").appendDescriptionOf(dataSourcesMatcher);
          ...
      }
    };
}
```

The above can either be in this file, or some 'Matchers' file somewhere.

You can argue this is more work, but the counter argument is that its reusable and makes the tests easier to read, understand, change and extend.